_posts/2017-09-21-scalamacros.md

```diff
@@ -0,0 +1,438 @@
+Scala REPL and other compiler plugins such as Scoverage.
+Most importantly however, we hit fundamental roadblockers when trying to add
+support for def macros with access to the semantic API.
+Nevertheless, we learned a few important lessons
+
+- Users and macro authors care a lot about IDE support, in particular IntelliJ
+  support.
+- Building a portable macro system with ASTs that don't leak compiler
+  implementation details is possible.
+- The Scalameta ASTs, which comprehensively cover all syntactic details
+  of Scala syntax, may be great to build developer tools such as scalafmt
+  but it ise too detailed for macro.
+  In particular, the detailed AST nodes introduced portability problems during
+  later phases of the compilation pipeline.
+
+With these lessons learned, we regretfully decided to retire our efforts to
+build a macro system on top of Scalameta.
+From now on, Scalameta's primary focus is to support building developer tools such
+as Scalafmt and Scalafix.
+
+### v3: scala.macros
+
+This spring, Eugene Burmako at Twitter and Liu Fengyun at EPFL worked on a
+new macro system to address the limitations of scala.reflect-based and
+scala.meta-based macros.
+The result of this multi-year effort is
+hosted at [scalamacros/scalamacros] and its design is explained in detail in the
+technical report
+["Two approaches to portable macros"](https://www.dropbox.com/s/2xzcczr3q77veg1/gestalt.pdf).
+This fourth iteration of Scala Macros builds on top of the strengths of the
+scala-reflect API with the following distinction, it's
+
+- smaller, the API exposes as little as possible from the compiler
+  while still being able to support most interesting macro applications.
+- more robust, with common pitfalls in scala.reflect-based macros guarded by
+  the type system with separation of typed and untyped trees.
+
+Scala Macros v3 build on top of a similar design as scala.reflect by providing
+an API to lazily extract and construct Scala syntax.
+It is no longer necessary to pass around path dependent trees or contexts,
+avoiding the need for macro bundles.
+Moreover, the def macro definition and implementation are merged into a single
+`def name = macro` definition, simplifying the macro expansion engine and
+preventing common mistakes made by novice macro authors.
+
+One paint point in macros v2 that macros v3 does not address is the separate
+compilation restriction.
+Macro definitions must still be compiled in a separate project from where they
+are used.
+Nevertheless, we believe that v3 represents a significant enough improvement to
+forgive this restriction, which is admittedly an inconvenience for macro author
+but arguably not a blocker for adoption.
+
+To give you a taste of the tentative macros v3 API, let's implement a `fieldNames`
+macro to extract fields names of a case class
+
+```scala
+import scala.macros._
+object CaseClass {
+  def fieldNames[T]: List[String] = macro {
+    val names = T.vals
```

Comment:
This is so confusing… Should be something `val names = typeOf[T].vals`. I hope you don’t plan to keep this magic in v4.